### DISCUSSION TITLE
Correct way to prevent type coercion for `datetime.datetime` objects

### DISCUSSION THREAD

**Discussion by figchutney** [pydantic/pydantic#2834] [Question] [answered]
Hey 👋 

I have a model that includes a `datetime.datetime` field:

```python
class Foo(pydantic.BaseModel):
    bar: datetime.datetime
```

For my particular use case, **I'd like to prevent coercion for `datetime.datetime` objects from happening completely**. I'm handling some pretty rough data, where I'd rather anything that isn't a `datetime.datetime` object results in `pydantic.ValidationError` being raised. Otherwise, I get some pretty whacky results: 

```zsh
>>> import pydantic
>>> import datetime

>>> class Foo(pydantic.BaseModel):
...     bar: datetime.datetime
...
>>> Foo(bar="20210101")
Foo(bar=datetime.datetime(1970, 8, 22, 21, 55, 1, tzinfo=datetime.timezone.utc))
```

`pydantic`'s [Strict Types](https://pydantic-docs.helpmanual.io/usage/types/#strict-types) seem to prevent coercion nicely for some types (`str`, `int`, `float` etc.), but I can't find an equivalent for `datetime.datetime`.

I had hoped using `pydantic.Field` with `strict=True` might work, but again assume this only applies to certain types.

What's the correct way of achieving the desired behaviour in this case? 🙏 I realise the primary aim of `pydantic` is _parsing_ and not _validating_, but assume there's a recommended way of doing this nonetheless :)

Thanks!

**Comment by PrettyWood** [comment, marked as answer]
Hi @figchutney 
You can
- add a pre-validator on `bar`
```py
from datetime import datetime

from pydantic import BaseModel, validator


class Foo(BaseModel):
    bar: datetime

    @validator('bar', pre=True)
    def only_instance(cls, v):
        if not isinstance(v, datetime):
            raise TypeError('v is not a datetime')
        return v


class Foo(BaseModel):
    bar: datetime
```
- create a custom type
```py
from datetime import datetime

from pydantic import BaseModel


class StrictDatetime(datetime):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        if not isinstance(v, datetime):
            raise TypeError('v is not a datetime')
        return v


class Foo(BaseModel):
    bar: StrictDatetime
```
- make your `datetime` strict with something like [this](https://github.com/samuelcolvin/pydantic/issues/2079#issuecomment-720118099). You won't need `typingx` in your case but the generic `Strict` type should get you there.

Hope it helps!

**Reply by figchutney** [reply]
Hey Eric 🙂  - thanks for taking the time to reply - a pre-validator works perfectly. Cheers! 👌 🚀
